src/coprocessor/dag/aggr_fn/impl_first.rs

```diff
@@ -0,0 +1,249 @@
+        max_columns: usize,
+        out_schema: &mut Vec<FieldType>,
+        out_exp: &mut Vec<RpnExpression>,
+    ) -> Result<Box<dyn super::AggrFunction>> {
+        use cop_datatype::FieldTypeAccessor;
+        use std::convert::TryFrom;
+        assert_eq!(aggr_def.get_tp(), ExprType::First);
+        let child = aggr_def.take_children().into_iter().next().unwrap();
+        let eval_type = EvalType::try_from(child.get_field_type().tp()).unwrap();
+
+        // FIRST outputs one column with the same type as its child
+        out_schema.push(aggr_def.take_field_type());
+
+        // FIRST doesn't need to cast, so using the expression directly.
+        out_exp.push(RpnExpressionBuilder::build_from_expr_tree(
+            child,
+            time_zone,
+            max_columns,
+        )?);
+
+        match_template_evaluable! {
+            TT, match eval_type {
+                EvalType::TT => Ok(Box::new(AggrFnFirst::<TT>::new()))
+            }
+        }
+    }
+}
+
+/// The FIRST aggregate function.
+#[derive(Debug, AggrFunction)]
+#[aggr_function(state = AggrFnStateFirst::<T>::new())]
+pub struct AggrFnFirst<T>(PhantomData<T>)
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>;
+
+impl<T> AggrFnFirst<T>
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>,
+{
+    fn new() -> Self {
+        AggrFnFirst(PhantomData)
+    }
+}
+
+/// The state of the FIRST aggregate function.
+#[derive(Debug)]
+pub enum AggrFnStateFirst<T>
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>,
+{
+    Empty,
+    Valued(Option<T>),
+}
+
+impl<T> AggrFnStateFirst<T>
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>,
+{
+    pub fn new() -> Self {
+        AggrFnStateFirst::Empty
+    }
+}
+
+// default impl all `Evaluable` for all `AggrFnStateFirst` to make `AggrFnStateFirst`
+// satisfy trait `AggrFunctionState`
+impl<T1, T2> super::AggrFunctionStateUpdatePartial<T1> for AggrFnStateFirst<T2>
+where
+    T1: Evaluable,
+    T2: Evaluable,
+    VectorValue: VectorValueExt<T2>,
+{
+    #[inline]
+    default fn update(&mut self, _ctx: &mut EvalContext, _value: &Option<T1>) -> Result<()> {
+        panic!("Unmatched parameter type")
+    }
+
+    #[inline]
+    default fn update_repeat(
+        &mut self,
+        _ctx: &mut EvalContext,
+        _value: &Option<T1>,
+        _repeat_times: usize,
+    ) -> Result<()> {
+        panic!("Unmatched parameter type")
+    }
+
+    #[inline]
+    default fn update_vector(
+        &mut self,
+        _ctx: &mut EvalContext,
+        _values: &[Option<T1>],
+    ) -> Result<()> {
+        panic!("Unmatched parameter type")
+    }
+}
+
+impl<T> super::AggrFunctionStateUpdatePartial<T> for AggrFnStateFirst<T>
+where
+    T: Evaluable,
+    VectorValue: VectorValueExt<T>,
+{
+    #[inline]
+    fn update(&mut self, _ctx: &mut EvalContext, value: &Option<T>) -> Result<()> {
+        if let AggrFnStateFirst::Empty = self {
+            // TODO: avoid this clone
```

Comment:
😕 There's no way to avoid the clone since value is an `&Option<T>` which cannot be moved out. You could avoid the clone only if value is `Option<T>` or `&mut Option<T>`.

Reply:
Yes, we cannot avoid it unless we change the parameter type. But we do not always need the clone, for example, we can actually move out the value if it is owned in the RpnStackNode. To accomplish that, we need a big change.

Reply:
But for an aggregate function, it's rare we need to clone the value. FIRST is quite a rare used aggregate function too. Maybe we can just ignore the problem here.

Reply:
Yes. We can ignore the clone cost, since it is called only once for every group, which is fine.